documentation/nat-service/nat-service.md

```diff
@@ -0,0 +1,206 @@
+
+* Both an outbound flow table entry and an inbound flow table entry are added to corresponding flow tables.
+
+* A search of the inbound flow table is triggered once the packet is re-injected. The packet will match the inbound flow table entry created previously and the flow action will be applied, which includes replacing the destination IP address/port number and adding VXLAN encapsulation.
+
+* A routing lookup is performed based on destination VTEP IP address before sending the VXLAN encapsulated packet to VM.
+
+### 2.2 Outbound packet processing pipeline
+
+![dnat-outbound-packet-processing-pipeline](images/Figure5-dnat-outbound-pipeline.png)
+*Figure 5 - DNAT outbound packet processing pipeline*
+
+* As described in section 2.1, the outbound flow table entry has been created during the handling of the *first packet* processing, when the responding packet arrives from VM, it matches the entry and performs the specified action of replacing the source IP address/port number and removing the VXLAN header before being sent to the internet.
+
+* Then it routes the packet based on DIP of the decapsulated packet.
+
+* Counting and/or metering is done, and they are optional based on the Vnet user's billing mode.
+
+● The packet is sent to Internet via the selected interface.
+
+## Configuration example
+
+``` JSON
+
+/* Define Vnet1 */
+DASH_VNET:Vnet1 {
+    "vni": 12345
+}
+
+/* Define Vnet2 */
+DASH_VNET:Vnet2 {
+    "vni": 45678
+}
+
+/* Define routing types */
+DASH_ROUTING_TYPE:vnet_encap: [
+    {
+         "name": "action1",
+         "action_type: "staticvxlan",
+         "encap_type" "vxlan"
+    }
+]
+
+/* Define Overlay routing tables */
+DASH_OVERLAY_ROUTING_TABLE:Vnet1:10.1.0.10 {
+    "routing_type":"vnet_encap", 
+    "dst_vtep":[30.0.0.22]
+    "mac_address":002244AABBCC
+},
+DASH_OVERLAY_ROUTING_TABLE:Vnet1:10.1.0.2 {
+    "routing_type":"vnet_encap", 
+    "dst_vtep":[30.0.0.23]
+    "mac_address":002244AABBDD
+}
+
+
+/* Define SNAT rules */
+DASH_SNAT_RULE_TABLE:Vnet1:10.1.0.0/24 {
+    "0.0.0.0/0": {
+        "src_ip_list":[111.1.1.237, 111.1.1.238]
+    }
+},
+DASH_SNAT_RULE_TABLE:Vnet1:10.1.1.0/24 {
+    "0.0.0.0/0": {
+        "src_ip_list":[111.1.1.239, 111.1.1.240]
+    }
+},
+DASH_SNAT_RULE_TABLE:Vnet2:10.1.0.0/24 {
+    "0.0.0.0/0": {
+        "src_ip_list":[111.1.1.241]
+    },
+    "102.0.0.0/8": {
+        "src_ip_list":[111.1.1.242]
+    },
+}
+
+/* Define DNAT rules */
+DASH_DNAT_RULE_TABLE:Vnet1:10.1.0.2:80 {
+    "mapping_list": [111.2.190.195:80, 114.66.253.4:8080]
+},
+DASH_DNAT_RULE_TABLE:Vnet1:10.1.0.3 {
+    "mapping_list": [111.2.190.196:80, 114.66.253.5]
+},
+
+/* Define NAT gateway vtep ip */
+DASH_NAT_GW_VTEP_TABLE{
+    "vtep_ip": [30.0.0.250]
+}
+```
+
+## Packet flow example
+
+### SNAT packet flow
+
+`Source: VM in Vnet1, VNI 12345, SIP 10.1.10.10`
+`Destination: IP 8.8.8.8, port 53, protocol UDP`
+
+* **first packet** processing flow of outbound packet
+
```

Comment:
What is the mechanism to pick a SIP (111.1.1.237 in this diagram) from a list of SIPs specified in the SNAT_RULE_TABLE?

Reply:
Prince asked this question in the meeting - Where did the translated UDP source port value of 45678 come from?

Reply:
For the populating of inbound flow entry, can dst_vtep and inner_mac_address be learned from the data packets outer_sip and inner_smac respectively?

Reply:
> What is the mechanism to pick a SIP (111.1.1.237 in this diagram) from a list of SIPs specified in the SNAT_RULE_TABLE?

The lookup takes 2 steps: 1) Use SIP + VNID to lookup into SNAT_RULE_TABLE( the SIP lookup should be LPM lookup); 2) Use DIP(also LPM) to lookup for a list of available SIP. After we get the list of SIP, use a load balancing algorithm to select SIP from the list.

Reply:
> Prince asked this question in the meeting - Where did the translated UDP source port value of 45678 come from?

Source port is allocated by the slowpath logic, we don't show it in the pipe for simplicity.

Reply:
> For the populating of inbound flow entry, can dst_vtep and inner_mac_address be learned from the data packets outer_sip and inner_smac respectively?

In my understanding, for DST_VTEP lookup,  we allow to configure a list of VTEP instead of learn it from the packet, for load balance purpose.
For Inner Dst MAC, learn from the packet is fine.

Reply:
> For the populating of inbound flow entry, can dst_vtep and inner_mac_address be learned from the data packets outer_sip and inner_smac respectively?

In my understanding, for DST_VTEP lookup,  we allow to configure a list of VTEP instead of learn it from the packet, for load balance purpose.
For Inner Dst MAC, learn from the packet is fine.

Reply:
> > Prince asked this question in the meeting - Where did the translated UDP source port value of 45678 come from?
> 
> Source port is allocated by the slowpath logic, we don't show it in the pipe for simplicity.

Are you planning to implement "slow path" in the P4 behavioral pipeline? If yes, we will need mechanisms for load-balancing to select a SIP and also SPORT in the P4 program.